modules/substrate/src/lib.rs

```diff
@@ -14,5 +14,254 @@
+		pub fn finalize_header(
+			origin,
+			hash: Hash<T::Header>,
+			finality_proof: Vec<u8>,
+		) -> dispatch::DispatchResult {
+			let _ = ensure_signed(origin)?;
+			frame_support::debug::trace!(target: "sub-bridge", "Got header hash {:?}", hash);
+
+			let mut verifier = verifier::Verifier {
+				storage: PalletStorage::<T>::new(),
+			};
+
+			let _ = verifier
+				.verify_finality(hash, &finality_proof)
+				.map_err(|_| <Error<T>>::UnfinalizedHeader)?;
+
+			Ok(())
+		}
+	}
+}
+
+/// Expected interface for interacting with bridge pallet storage.
+pub trait BridgeStorage {
+	/// The header type being used by the pallet.
+	type Header: HeaderT;
+
+	/// Write a header to storage.
+	fn write_header(&mut self, header: &ImportedHeader<Self::Header>);
+
+	/// Get the best finalized header the pallet knows of.
+	fn best_finalized_header(&self) -> ImportedHeader<Self::Header>;
+
+	/// Update the best finalized header the pallet knows of.
+	fn update_best_finalized(&self, hash: <Self::Header as HeaderT>::Hash);
+
+	/// Check if a particular header is known to the pallet.
+	fn header_exists(&self, hash: <Self::Header as HeaderT>::Hash) -> bool;
+
+	/// Get a specific header by its hash.
+	///
+	/// Returns None if it is not known to the pallet.
+	fn header_by_hash(&self, hash: <Self::Header as HeaderT>::Hash) -> Option<ImportedHeader<Self::Header>>;
+
+	/// Get the current Grandpa authority set.
+	fn current_authority_set(&self) -> AuthoritySet;
+
+	/// Update the current Grandpa authority set.
+	///
+	/// Should only be updated when a scheduled change has been triggered.
+	fn update_current_authority_set(&self, new_set: AuthoritySet);
+
+	/// Replace the current authority set with the next scheduled set.
+	fn enact_authority_set(&mut self);
+
+	/// Get the next scheduled Grandpa authority set change.
+	fn scheduled_set_change(&self) -> Option<ScheduledChange<<Self::Header as HeaderT>::Number>>;
+
+	/// Schedule a Grandpa authority set change in the future.
+	fn schedule_next_set_change(&self, next_change: ScheduledChange<<Self::Header as HeaderT>::Number>);
+}
+
+/// Used to interact with the pallet storage in a more abstract way.
+#[derive(Default, Clone)]
+pub struct PalletStorage<T>(PhantomData<T>);
+
+impl<T> PalletStorage<T> {
+	fn new() -> Self {
+		Self(PhantomData::<T>::default())
+	}
+}
+
+impl<T: Trait> BridgeStorage for PalletStorage<T> {
+	type Header = T::Header;
+
+	fn write_header(&mut self, header: &ImportedHeader<T::Header>) {
+		let hash = header.header.hash();
+		<ImportedHeaders<T>>::insert(hash, header);
+	}
+
+	fn best_finalized_header(&self) -> ImportedHeader<T::Header> {
+		let hash = <BestFinalized<T>>::get();
+		self.header_by_hash(hash)
+			.expect("A finalized header was added at genesis, therefore this must always exist")
+	}
+
+	fn update_best_finalized(&self, hash: Hash<T::Header>) {
+		<BestFinalized<T>>::put(hash)
+	}
+
+	fn header_exists(&self, hash: Hash<T::Header>) -> bool {
+		self.header_by_hash(hash).is_some()
```

Comment:
It was already reported earlier by Slava that this should rather use `contains_key` to avoid decoding the header.
This could be left as a default implementation on the trait level instead.